Code/GraphMol/ChemReactions/Enumerate/Enumerate.cpp

```diff
@@ -0,0 +1,189 @@
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Novartis Institutes for BioMedical Research Inc.
+//       nor the names of its contributors may be used to endorse or promote
+//       products derived from this software without specific prior written
+//       permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "Enumerate.h"
+#include "CartesianProduct.h"
+#include "RandomSample.h"
+#include "RandomSampleAllBBs.h"
+#include "EvenSamplePairs.h"
+#include "../ReactionPickler.h"
+#include <GraphMol/MolPickler.h>
+#include <GraphMol/SmilesParse/SmilesWrite.h>
+#include "EnumerationPickler.h"
+
+// Since we are exporting the classes for serialization,
+//  we should declare the archives types used here
+#include <RDGeneral/BoostStartInclude.h>
+#include <boost/archive/binary_oarchive.hpp>
+#include <boost/archive/binary_iarchive.hpp>
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
+#include <boost/serialization/shared_ptr.hpp>
+#include <boost/serialization/export.hpp>
+#include <RDGeneral/BoostEndInclude.h>
+
+BOOST_CLASS_EXPORT(RDKit::EnumerationStrategyBase);
+BOOST_CLASS_EXPORT(RDKit::CartesianProductStrategy);
+BOOST_CLASS_EXPORT(RDKit::RandomSampleStrategy);
+BOOST_CLASS_EXPORT(RDKit::RandomSampleAllBBsStrategy);
+BOOST_CLASS_EXPORT(RDKit::EvenSamplePairsStrategy);
+BOOST_CLASS_EXPORT(RDKit::EnumerateLibrary);
+
+namespace RDKit {
+const RGROUPS &EnumerateLibraryBase::getPosition() const {
+  return m_enumerator->getPosition();
+}
+
+std::string EnumerateLibraryBase::getState() const {
+  PRECONDITION(m_enumerator.get(), "Null Enumerator");
+  std::string state;
+  EnumerationStrategyPickler::pickle(m_enumerator, state);
+  return state;
+}
+
+void EnumerateLibraryBase::setState(const std::string &state) {
+  m_enumerator = EnumerationStrategyPickler::fromPickle(state);
+}
+
+void EnumerateLibraryBase::resetState() {
+  setState(m_initialState);
+}
+
+std::vector<std::vector<std::string> > EnumerateLibraryBase::nextSmiles() {
```

Comment:
I believe this is return value optimized optimized by copy elision.

http://en.cppreference.com/w/cpp/language/copy_elision

It can also be a move constructor in c++11.  In any-case if there is a copy, it is really low cost for standard returns and will not be a copy in c++11 and 14.